File: data_structures/binary_tree/BinTree.cs
using System;

namespace BinaryTree {
    
    public class Node 
    {
        public int data { get; set; }
        public Node Left { get; set; }
        public Node Right { get; set; }

        public Node(int data)
        {
            this.data = data;
            this.Left = null;
            this.Right = null;
        }
    }

    public class BinaryTree {
        
        public Node Root { get; }
        public Node SortedRoot { get; }
        public BinaryTree(int[] values) 
        {
            this.Root = this.BuildTree(this.Root, values, 0);
            this.SortedRoot = this.BuildTreeSorted(this.SortedRoot, values);
        }

        private Node BuildTree(Node node, int[] values, int itr) 
        {
            if (itr < values.Length)
            {   
                node = new Node(values[itr]);
                node.Left = this.BuildTree(node.Left, values, itr * 2 + 1);
                node.Right = this.BuildTree(node.Right, values, itr * 2 + 2);
            }
            return node;
        }

        private Node Insert(Node node, int value)
        {
            if (node == null)
            {
                return new Node(value);
            }
            if (value <= node.data)
            {
                node.Left = this.Insert(node.Left, value);
            }
            else if (value > node.data) 
            {
                node.Right = this.Insert(node.Right, value);
            }
            return node;
        }

        private Node BuildTreeSorted(Node Root, int[] values)
        {
             Root = null;
             Root = this.Insert(Root, values[0]);
             for(int i = 1; i < values.Length; i++)
             {
                 this.Insert(Root, values[i]);
             }
             return Root;
        }

        public void PreorderApply(Node node, Del<Node> Callback)
        {
            if (node != null)
            {
                Callback(node);
                this.PreorderApply(node.Left, Callback);
                this.PreorderApply(node.Right, Callback);
            }
        }

        public void InorderApply(Node node, Del<Node> Callback)
        {
            if (node != null)
            {
                this.InorderApply(node.Left, Callback);
                Callback(node);
                this.InorderApply(node.Right, Callback);
            }
        }

        public void PostOrderApply(Node node, Del<Node> Callback)
        {
            if (node != null)
            {
                this.PostOrderApply(node.Left, Callback);
                this.PostOrderApply(node.Right, Callback);
                Callback(node);
            }
        }
    }

    public delegate void Del<T>(T param);

    public class Program
    {

        public static void PrintNodeValue(Node n)
        {
            Console.Write($"{n.data}, ");
        }

        public static void Main(string[] args)
        {
            int[] values = {1,2,3,4,5};
            var Btree = new BinaryTree(values);
            Del<Node> Callback = Program.PrintNodeValue;

            Btree.PreorderApply(Btree.Root, Callback);
            Console.WriteLine();
            Btree.InorderApply(Btree.SortedRoot, Callback);
            Console.WriteLine();
            Btree.PostOrderApply(Btree.Root, Callback);
        }
    }
}

File: data_structures/binary_tree/c/bin_tree.c
#include "bin_tree.h"
#include "helper.c"

int main()
{
    int size = 5;
    int unsorted_values[5] = {22,3,1,66,4};
    Node *sorted_root = sorted_binary_tree(unsorted_values, size);
    
    preorder_apply(sorted_root, print_value);
    printf("\n");
    inorder_apply(sorted_root, print_value);
    printf("\n");
    postorder_apply(sorted_root, print_value);

    return (0);
}

Node *insert(Node *node, int value)
{
    if (node == NULL)
    {
        return new_node(value);
    }
    if (value <= node->data)
    {
        node->left = insert(node->left, value);
    }
    else if (value > node->data)
    {
        node->right = insert(node->right, value);
    }
    return node;
}

Node *sorted_binary_tree(int *values, int size)
{
    Node *root = NULL; 
    root = insert(root, values[0]);

    for (int i = 1; i < size; i++ )
    {
        insert(root, values[i]);
    }
    return root;
}

Node *binary_tree(Node *root, int *values, int i, int size)
{
    if (i < size)
    {
        root = new_node(values[i]);
        root->left = binary_tree(root->left, values, 2 * i + 1, size);
        root->right = binary_tree(root->right, values, 2 * i + 2, size);
    }
    return root;
}

Node *new_node(int data)
{
    Node *new_node = (Node *)malloc(sizeof(Node));
    new_node->data = data;
    new_node->left = NULL;
    new_node->right = NULL;
    return (new_node);
}

void preorder_apply(Node *root, void (*fn)(Node *))
{
    if (!root) {
        return;
    }

    (*fn)(root);
    preorder_apply(root->left, fn);
    preorder_apply(root->right, fn);
}

void inorder_apply(Node *root, void (*fn)(Node *))
{
    if(!root) { 
        return; 
    }
 
    inorder_apply(root->left, fn); 

    (*fn)(root);

    inorder_apply(root->right, fn);
}

void postorder_apply(Node *root, void (*fn)(Node *)) 
{
    if(!root)
    {
        return;
    }

    postorder_apply(root->left, fn);
    postorder_apply(root->right, fn);
    (*fn)(root);
}


